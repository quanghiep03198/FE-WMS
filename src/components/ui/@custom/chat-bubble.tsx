import { cn } from '@/common/utils/cn'
import { cva, VariantProps } from 'class-variance-authority'
import React from 'react'

const chatBubbleVariant = cva('rounded-lg max-w-64 text-sm p-3', {
	variants: {
		variant: {
			primary: 'bg-primary text-primary-foreground',
			secondary: 'bg-secondary text-secondary-foreground'
		},
		side: {
			left: '[--_p:_100%] !rounded-br-[0_0]',
			right: '[--_p:_0] !rounded-bl-[0_0]'
		}
	}
})

export interface ChatBubbleProps
	extends React.PropsWithChildren,
		React.ComponentProps<'div'>,
		VariantProps<typeof chatBubbleVariant> {}

const ChatBubble: React.FC<ChatBubbleProps> = ({ children, variant, side, className }) => {
	return (
		<div
			className={cn(chatBubbleVariant({ variant, side, className }))}
			style={{
				...({
					'--_p': side === 'left' ? '100%' : '0' /* the position of the tail */,
					'--r': '0.5em' /* the radius */,
					'--t': '1.25em' /* the size of the tail */
				} as React.CSSProperties),
				borderRadius: 'calc(var(--r) + var(--t)) / var(--r)',
				borderInline: 'var(--t) solid transparent',
				mask: 'radial-gradient(100% 100% at var(--_p) 0, transparent 100%, black 105%) var(--_p) 100% / var(--t) var(--t) no-repeat, linear-gradient(black 0 0) padding-box'
			}}>
			{children}
		</div>
	)
}

export default ChatBubble
