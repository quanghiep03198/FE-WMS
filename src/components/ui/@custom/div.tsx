import { cn } from '@/common/utils/cn';
import React, { forwardRef, useRef } from 'react';

type TBoxProps = { as?: React.ElementType } & React.PropsWithChildren & React.ComponentProps<React.ElementType>;

export const Div: React.ForwardRefExoticComponent<TBoxProps> = forwardRef((props, ref) => {
	const { as: Component = 'div', className, style, children, ...restProps } = props;

	const localRef = useRef(null);
	const resolvedRef = ref || localRef;

	return (
		<Component className={cn(className)} style={style} ref={resolvedRef} {...restProps}>
			{children}
		</Component>
	);
});
