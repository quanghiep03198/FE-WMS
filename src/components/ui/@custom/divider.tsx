import { PropsWithChildren } from 'react';

export const Divider: React.FC<PropsWithChildren> = ({ children }) => {
	return (
		<div className='relative'>
			<div className='absolute inset-0 flex items-center' aria-hidden='true'>
				<div className='w-full border-t' />
			</div>
			<div className='relative flex justify-center text-sm font-medium leading-6'>
				<span className='bg-background px-6 text-xs uppercase text-muted-foreground'>{children}</span>
			</div>
		</div>
	);
};
