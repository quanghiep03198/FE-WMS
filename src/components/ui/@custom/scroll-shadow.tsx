import { cn } from '@/common/utils/cn'
import { useScroll, useSize, useUpdate } from 'ahooks'
import React, { forwardRef, memo, useEffect, useRef } from 'react'

export interface ScrollShadowProps extends React.PropsWithChildren, React.ComponentProps<'div'> {
	/**
	 * If size is a number, it will be treated as pixels, otherwise is a string, it will be treated as a CSS value such as px, rem or em.
	 */
	size?: React.CSSProperties['height']
	scrollbar?: boolean
	ref?: React.Ref<HTMLDivElement>
}

const ScrollShadow = forwardRef<HTMLDivElement, ScrollShadowProps>(
	({ size = 320, scrollbar = false, className, children }, ref) => {
		const update = useUpdate()
		const localContainerRef = useRef<HTMLDivElement>(null)
		const resolvedRef = (ref ?? localContainerRef) as React.MutableRefObject<HTMLDivElement>

		const containerScroll = useScroll(resolvedRef)
		const clientSize = useSize(resolvedRef)

		const scrollHeight = resolvedRef.current?.scrollHeight ?? 0
		const scrollTop = resolvedRef.current?.scrollTop ?? 0
		const scrollClientHeight = resolvedRef.current?.clientHeight ?? 0

		const isScrolledToTop = containerScroll?.top === 0
		const isScrolledToBottom = scrollHeight - scrollTop - scrollClientHeight < 1
		const isScrollTopBottom = !isScrolledToTop && !isScrolledToBottom

		const isScrollable = scrollHeight > scrollClientHeight

		useEffect(update, [scrollHeight, scrollTop, scrollClientHeight, clientSize])

		return (
			<div
				ref={resolvedRef}
				style={{ '--scroll-shadow-size': typeof size === 'number' ? `${size}px` : size } as React.CSSProperties}
				data-top-scroll={isScrollable && isScrolledToTop}
				data-bottom-scroll={isScrollable && isScrolledToBottom}
				data-top-bottom-scroll={isScrollable && isScrollTopBottom}
				className={cn(
					className,
					scrollbar ? 'overflow-y-scroll scrollbar' : 'overflow-auto !scrollbar-none',
					'h-[var(--scroll-shadow-size)]',
					'data-[bottom-scroll=true]:[mask-image:linear-gradient(0deg,hsl(var(--sidebar-background))_calc(100%_-_var(--scroll-shadow-size)/4),transparent)]',
					'data-[top-scroll=true]:[mask-image:linear-gradient(180deg,hsl(var(--sidebar-background))_calc(100%_-_var(--scroll-shadow-size)/4),transparent)]',
					'data-[top-bottom-scroll=true]:[mask-image:linear-gradient(180deg,transparent,hsl(var(--sidebar-background))_calc(var(--scroll-shadow-size)/4),hsl(var(--sidebar-background))_calc(100%_-_var(--scroll-shadow-size)/4),transparent)]'
				)}>
				{children}
			</div>
		)
	}
)

export default memo(ScrollShadow)
