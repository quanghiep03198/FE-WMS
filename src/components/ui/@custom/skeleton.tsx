import { cn } from '@/common/utils/cn'
import React from 'react'

const Skeleton: React.FC<React.ComponentProps<'div'>> = ({ className, ...props }) => {
	return (
		<div
			{...props}
			className={cn(
				'min-h-3 w-full animate-shimmer rounded-[2px] bg-[linear-gradient(120deg,hsl(var(--secondary)),45%,hsl(var(--background)),55%,hsl(var(--secondary)))] bg-[length:200%_100%] dark:bg-[linear-gradient(120deg,hsl(var(--muted)),45%,hsl(var(--muted-foreground)),55%,hsl(var(--muted)))]',
				className
			)}
		/>
	)
}

export default Skeleton
