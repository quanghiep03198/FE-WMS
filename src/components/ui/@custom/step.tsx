import { Div, Icon, Typography } from '@/components/ui'
import { TFunction } from 'i18next'
import React, { createContext, useContext, useReducer } from 'react'
import { useTranslation } from 'react-i18next'
import tw from 'tailwind-styled-components'

// #region Step types
export type TStep = {
	index?: number
	name: FirstParameter<TFunction<undefined, undefined>>
	status: 'current' | 'upcoming' | 'completed'
}

type TStepState = {
	data: TStep[]
	finishedSteps: number[]
	currentStep: number
	canNextStep: boolean
	canPrevStep: boolean
}

type TStepAction =
	| { type: 'PREV_STEP' }
	| { type: 'NEXT_STEP' }
	| { type: 'GO_TO_STEP'; payload: number }
	| { type: 'COMPLETE' }

type TStepContext = {
	steps: TStepState
	dispatch: React.Dispatch<TStepAction>
}
// #endregion

//	#region Step context
const reducer: React.Reducer<TStepState, TStepAction> = (state, action) => {
	switch (action.type) {
		case 'PREV_STEP': {
			if (!state.canPrevStep) return state
			return {
				...state,
				currentStep: state.currentStep - 1,
				data: state.data.map((step) => {
					const status =
						step.index! < state.currentStep - 1
							? 'completed'
							: step.index === state.currentStep - 1
								? 'current'
								: 'upcoming'
					return { ...step, status }
				})
			}
		}

		case 'NEXT_STEP': {
			const currentStep = state.currentStep + 1
			if (state.currentStep == state.data.length) return state

			return {
				...state,
				currentStep: currentStep,
				canNextStep: currentStep < state.data.length && state.finishedSteps.includes(state.currentStep),
				canPrevStep: true,
				finishedSteps: [...new Set([state.currentStep, ...state.finishedSteps])].sort((a, b) => a - b),
				data: state.data.map((step) => {
					const status =
						step.index! < currentStep ? 'completed' : step.index === currentStep ? 'current' : 'upcoming'
					return { ...step, status }
				})
			}
		}

		case 'GO_TO_STEP': {
			const currentStep = action.payload

			return {
				...state,
				currentStep: currentStep,
				canNextStep: currentStep < state.data.length && state.finishedSteps.includes(state.currentStep),
				// canPrevStep: state.finishedSteps.some((step) => currentStep >= step),
				data: state.data.map((step) => {
					const status =
						step.index < currentStep ? 'completed' : step.index === currentStep ? 'current' : 'upcoming'
					return { ...step, status }
				})
			}
		}

		case 'COMPLETE': {
			return {
				...state,
				currentStep: state.data.length,
				canNextStep: false,
				canPrevStep: true,
				data: state.data.map((step) => ({ ...step, status: 'completed' })),
				finishedSteps: [...new Set([...state.finishedSteps, state.data.length])].sort((a, b) => a - b)
			}
		}

		default: {
			throw new Error('Invalid step action')
		}
	}
}
const initialState: TStepState = { data: [], finishedSteps: [], currentStep: 1, canNextStep: false, canPrevStep: false }

const StepContext = createContext<TStepContext>(null)

const StepProvider: React.FC<{ data: TStep[]; enableChangeStep?: boolean } & React.PropsWithChildren> = ({
	data,
	enableChangeStep = true,
	children
}) => {
	const [steps, dispatch] = useReducer(reducer, {
		...initialState,
		data: data.map((step: TStep, index: number) => ({ ...step, index: index + 1 }))
	})

	return (
		<StepContext.Provider value={{ steps, dispatch }}>
			<Div className='space-y-6'>
				<Steps enableChangeStep={enableChangeStep} />
				{children}
			</Div>
		</StepContext.Provider>
	)
}
// #endregion

// #region Step components
const StepList = tw.ol`grid grid-flow-col sm:grid-flow-row auto-cols-fr isolate divide-y-0 divide-border rounded-md border sm:grid-cols-1 sm:divide-y transition-all`
const Step = tw.li`flex items-center py-6 text-sm relative hover:opacity-80 px-4 whitespace-nowrap font-medium`
const StepContent = tw.div`inline-flex items-center justify-center gap-x-4 px-4`
const StepIndex = tw.div`inline-grid size-9 place-content-center rounded-full border-2 font-semibold`

const Steps: React.FC<{ enableChangeStep?: boolean }> = ({ enableChangeStep }) => {
	const { t } = useTranslation()

	const {
		steps: { data, finishedSteps },
		dispatch
	} = useContext(StepContext)

	const handleChangeStep = (step: TStep) => {
		if (!enableChangeStep) return
		if (finishedSteps.includes(step.index) || step.index === finishedSteps.at(-1) + 1) {
			dispatch({ type: 'GO_TO_STEP', payload: step.index })
		}
	}

	return (
		<Div as='nav' aria-label='Progress' className='w-full bg-background'>
			<StepList role='list'>
				{data.map((step: TStep) => {
					const stepName = t(step.name, { defaultValue: step.name, ns: undefined })

					return (
						<Step key={step.index} onClick={() => handleChangeStep(step)}>
							{step.status === 'completed' ? (
								<StepContent>
									<StepIndex className='border-success bg-success duration-200 group-hover:bg-success/80'>
										<Icon name='Check' size={20} className='text-success-foreground' aria-hidden='true' />
									</StepIndex>
									<Typography variant='small' color='success' className='font-medium'>
										{stepName}
									</Typography>
								</StepContent>
							) : step.status === 'current' ? (
								<StepContent>
									<StepIndex className='border-success text-success'>{step.index}</StepIndex>
									<Typography variant='small' color='success'>
										{stepName}
									</Typography>
								</StepContent>
							) : (
								<StepContent>
									<StepIndex className='text-muted-foreground'>{step.index}</StepIndex>
									<Typography variant='small' color='muted'>
										{stepName}
									</Typography>
								</StepContent>
							)}
							{step.index !== data.length && <StepSeparator />}
						</Step>
					)
				})}
			</StepList>
		</Div>
	)
}

const StepSeparator: React.FC = () => (
	<div className='absolute right-0 top-0 h-full w-5 translate-x-1/2 sm:hidden' aria-hidden='true'>
		<svg className='h-full w-full text-border' viewBox='0 0 22 80' fill='none' preserveAspectRatio='none'>
			<path d='M0 -2L20 40L0 82' vectorEffect='non-scaling-stroke' stroke='currentcolor' strokeLinejoin='round' />
		</svg>
	</div>
)

const StepPanel: React.FC<{ value: TStep['index'] } & React.PropsWithChildren> = ({ value, children }) => {
	const {
		steps: { currentStep }
	} = useStepContext()

	return currentStep === value && children
}
// #endregion

export const useStepContext = () => useContext(StepContext)

export const Stepper = {
	Panel: StepPanel,
	Provider: StepProvider
}
