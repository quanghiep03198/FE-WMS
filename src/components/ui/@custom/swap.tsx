import { forwardRef, useId, useRef } from 'react';
import tw from 'tailwind-styled-components';
import { Button } from '..';

type TSwapButtonProps = React.ClassAttributes<HTMLLabelElement> &
	React.LabelHTMLAttributes<HTMLLabelElement> & {
		checked?: boolean;
		onChange?: React.ChangeEventHandler<HTMLInputElement>;
		swapOn: React.ReactNode;
		swapOff: React.ReactNode;
	};

export const Swap = forwardRef<React.Ref<HTMLInputElement>, TSwapButtonProps>(({ checked, swapOn, swapOff, onChange, ...rest }: TSwapButtonProps, ref) => {
	const id = useId();
	const localRef = useRef(null);
	const resolvedRef = (ref || localRef) as React.MutableRefObject<any>;

	return (
		<Label htmlFor={id} {...rest}>
			<Toggler
				id={id}
				type='checkbox'
				checked={checked}
				ref={resolvedRef}
				onChange={(e) => {
					if (onChange) onChange(e);
				}}
			/>
			<SwapLabel className='opacity-0 peer-checked:-rotate-45 peer-checked:opacity-100 peer-indeterminate:opacity-0'>{swapOn}</SwapLabel>
			<SwapLabel className='opacity-100 peer-checked:-rotate-45 peer-checked:opacity-0'>{swapOff}</SwapLabel>
		</Label>
	);
});
const Label = tw.label`
	inline-grid 
	items-center
	justify-center
	whitespace-nowrap
	rounded-md
	text-sm
	font-medium
	focus-visible:outline-none
	focus-visible:ring-1
	focus-visible:ring-ring
	disabled:pointer-events-none
	disabled:opacity-50
	place-content-center
	cursor-pointer
	relative
	select-none
	padding-1
	hover:bg-accent
	hover:text-accent-foreground
	h-9 w-9


`;

const Toggler = tw.input`
	peer
	appearance-none
	absolute
	inset-0
`;
const SwapLabel = tw.div`
	cursor-pointer
	select-none
	text-base
	row-start-1
	col-start-1
	align-middle
	transition-all 
	duration-300
`;
