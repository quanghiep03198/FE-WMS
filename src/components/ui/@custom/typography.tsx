import { cn } from '@/common/utils/cn'
import { cva, type VariantProps } from 'class-variance-authority'
import { forwardRef, useRef } from 'react'

type TypographyProps = {
	as?: React.ElementType
} & React.HTMLAttributes<HTMLElement> &
	VariantProps<typeof typographyVariants> &
	React.PropsWithChildren

export const typographyVariants = cva('', {
	variants: {
		variant: {
			default: 'text-base tracking-tight',
			h1: 'text-6xl sm:text-5xl scroll-m-20 font-bold tracking-tight',
			h2: 'text-5xl sm:text-4xl font-bold scroll-m-20 tracking-tight',
			h3: 'text-4xl sm:text-3xl font-bold tracking-tight scroll-m-20',
			h4: 'text-3xl sm:text-2xl font-semibold tracking-tight scroll-m-20',
			h5: 'text-2xl sm:text-xl font-semibold tracking-tight scroll-m-20',
			h6: 'text-xl sm:text-lg font-semibold',
			p: 'leading-7',
			code: 'leading-7 font-mono',
			blockquote: 'mt-6 border-l-2 pl-6 italic',
			small: 'text-sm leading-snug'
		},
		color: {
			default: 'text-[inherit]',
			primary: 'text-primary',
			accent: 'accent',
			secondary: 'text-secondary',
			muted: 'text-muted-foreground',
			success: 'text-success',
			destructive: 'text-destructive'
		}
	},
	defaultVariants: {
		variant: 'default',
		color: 'default'
	}
})

export const Typography = forwardRef<HTMLElement, TypographyProps>((props, ref) => {
	const { as = 'p', className, children, color, variant, ...restProps } = props

	const localRef = useRef(null)
	const resolvedRef = ref ?? localRef

	const Element = !variant || variant === 'default' || as ? as : (variant as React.ElementType)

	return (
		<Element ref={resolvedRef} className={cn(typographyVariants({ variant, color, className }))} {...restProps}>
			{children}
		</Element>
	)
})
