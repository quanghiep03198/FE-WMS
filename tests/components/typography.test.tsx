import { Typography } from '@/components/ui/@custom/typography'
import { render } from '@testing-library/react'

describe('Typography Component', () => {
	it('renders correctly with default props', () => {
		const { container } = render(<Typography>Default Text</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with primary color', () => {
		const { container } = render(<Typography color='primary'>Primary Text</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with custom element', () => {
		const { container } = render(<Typography as='span'>Span Text</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with combined variant and color', () => {
		const { container } = render(
			<Typography variant='h2' color='accent'>
				Accent Heading 2
			</Typography>
		)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with success color', () => {
		const { container } = render(<Typography color='success'>Success Text</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with destructive color', () => {
		const { container } = render(<Typography color='destructive'>Destructive Text</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with h1 variant', () => {
		const { container } = render(<Typography variant='h1'>Heading 1</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with h2 variant', () => {
		const { container } = render(<Typography variant='h2'>Heading 2</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with h3 variant', () => {
		const { container } = render(<Typography variant='h3'>Heading 3</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly with h4 variant', () => {
		const { container } = render(<Typography variant='h4'>Heading 4</Typography>)
		expect(container.firstChild).toMatchSnapshot()
	})
})
