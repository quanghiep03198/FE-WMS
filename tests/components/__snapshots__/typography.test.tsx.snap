// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography Component > renders correctly with combined variant and color 1`] = `
<p
  class="text-5xl sm:text-4xl font-bold scroll-m-20 tracking-tight text-accent"
>
  Accent Heading 2
</p>
`;

exports[`Typography Component > renders correctly with custom element 1`] = `
<span
  class="text-base tracking-tight text-[inherit]"
>
  Span Text
</span>
`;

exports[`Typography Component > renders correctly with default props 1`] = `
<p
  class="text-base tracking-tight text-[inherit]"
>
  Default Text
</p>
`;

exports[`Typography Component > renders correctly with destructive color 1`] = `
<p
  class="text-base tracking-tight text-destructive"
>
  Destructive Text
</p>
`;

exports[`Typography Component > renders correctly with h1 variant 1`] = `
<p
  class="text-6xl sm:text-5xl scroll-m-20 font-bold tracking-tight text-[inherit]"
>
  Heading 1
</p>
`;

exports[`Typography Component > renders correctly with h2 variant 1`] = `
<p
  class="text-5xl sm:text-4xl font-bold scroll-m-20 tracking-tight text-[inherit]"
>
  Heading 2
</p>
`;

exports[`Typography Component > renders correctly with h3 variant 1`] = `
<p
  class="text-4xl sm:text-3xl font-bold tracking-tight scroll-m-20 text-[inherit]"
>
  Heading 3
</p>
`;

exports[`Typography Component > renders correctly with h4 variant 1`] = `
<p
  class="text-3xl sm:text-2xl font-semibold tracking-tight scroll-m-20 text-[inherit]"
>
  Heading 4
</p>
`;

exports[`Typography Component > renders correctly with primary color 1`] = `
<p
  class="text-base tracking-tight text-primary"
>
  Primary Text
</p>
`;

exports[`Typography Component > renders correctly with success color 1`] = `
<p
  class="text-base tracking-tight text-success"
>
  Success Text
</p>
`;
